### PR TITLE
Support Webpack 5, drop copypasted webpack-virtual-modules

### LINKED ISSUE #126: Issue with webpack cache-loader。
When i build my svelte app by webpack with `cache-loader` introduced, i've got an error:
```
typeError: stats.mtime.getTime is not a function
    at FS.stat (/Users/xxx/svelte-webpack/node_modules/cache-loader/dist/index.js:114:33)
    at process.nextTick (/Users/xxx/svelte-webpack/node_modules/enhanced-resolve/lib/CachedInputFileSystem.js:85:15)
    at process._tickCallback (internal/process/next_tick.js:61:11)
    ...
```

And then i print out the `stats` object into the terminal to find out what the 'mtime' key is, the logs are as below:

```
log => Stats {
  dev: 16777220,
  mode: 33188,
  nlink: 1,
  uid: 501,
  gid: 20,
  rdev: 0,
  blksize: 4194304,
  ino: 15306205,
  size: 51,
  blocks: 8,
  atimeMs: 1593482967553.3503,
  mtimeMs: 499162500000,
  ctimeMs: 1592617381679.1177,
  birthtimeMs: 499162500000,
  atime: 2020-06-30T02:09:27.553Z,
  mtime: 1985-10-26T08:15:00.000Z,
  ctime: 2020-06-20T01:43:01.679Z,
  birthtime: 1985-10-26T08:15:00.000Z }

log => VirtualStats {
  dev: 8675309,
  nlink: 0,
  uid: 1000,
  gid: 1000,
  rdev: 0,
  blksize: 4096,
  ino: 45000000,
  mode: 33188,
  size: 1219,
  blocks: 0,
  atime: 1593483324657,
  mtime: 1593483324657,
  ctime: 1593483324657,
  birthtime: 1593483324657 }
```

According to these, we can see the `mtime` is not a date object in `VirtualStats` which is provided in `svelte-loader/lib/virtual.js`:

```
// Starts in line 49.
VirtualModulesPlugin.prototype.writeModule = function(filePath, contents) {
    var len = contents ? contents.length : 0;
    var time = Date.now(); // Here it is, it's just a time stamp, not a date object.

    var stats = new VirtualStats({
        // ...
        mtime: time,
        // ...
    });
    // ...
};
```

But webpack `cache-loader` treats the `mtime` as a date object, and will invoke its `getTime` method, that's the problem!

### LINKED ISSUE #131: Component css gets reloaded with css-loader on any file change
Hi dear devs and contributors.

I've noticed this behavior trying to reduce watch mode reload times when using tailwind+postcss through svelte-preprocess by monitoring with speed monitoring webpack plugin.

On any non-component change in the project with 20 components that uses in-component css there's 40 seconds of wait due to reloading of all `Component.svelte.css` in `_virtualFiles`.

I've been messing around with `virtual.js` for some hours trying to fix this.

Tried to pass `fs.statSync(this.resourcePath)`'s to use with virtual css file. Though that non-authentic VirtualFileStats triggered watcher to reload. Didn't work.

console.logged contents of `_statStorage`, `_readFileStorage`, `_virtualFiles` in `compiler.inputFileSystem.purge` and `compiler.inputFileSystem._writeVirtualFile`.
It appears that original purge doesn't clear any of that containers, commenting it's code doesn't change anything.
`compiler.inputFileSystem.purge` does trigger on any change.
https://github.com/sveltejs/svelte-loader/blob/929eebe391597528d5a485e0563d82cee047686c/lib/virtual.js#L15-L30

While we at it, what is the purpose of this?
https://github.com/sveltejs/svelte-loader/blob/929eebe391597528d5a485e0563d82cee047686c/lib/virtual.js#L40-L49

Also there's an issue with similar task and problem, where author solved it by decorating `NodeWatchFileSystem`.
https://github.com/webpack/webpack/issues/5824
Don't know if that is a good idea...

The main question is why watcher reloads virtual css files when they weren't changed...
Would like to get some insight into this.

### LINKED ISSUE #139: Webpack 5. Error: TypeError: TypeError: Cannot read property 'data' of undefined
im in the process of updating all the module federation repos to the final v5

Ive run into issues with svelt-loader, but know very little about svelte..

This is the branch and specific project im trying to run this against:
https://github.com/module-federation/module-federation-examples/blob/update-to-v5-stable/comprehensive-demo/app-04/webpack.config.js

```
[0] @comprehensive-demo/app-04: ERROR in ./src/App.svelte
[0] @comprehensive-demo/app-04: Module build failed (from ./node_modules/svelte-loader/index.js):
[0] @comprehensive-demo/app-04: Error: TypeError: TypeError: Cannot read property 'data' of undefined
[0] @comprehensive-demo/app-04:     at /Volumes/lulu_dev/module-federation-examples/comprehensive-demo/app-04/node_modules/svelte-loader/index.js:181:12
[0] @comprehensive-demo/app-04:  @ ./src/main.js 1:0-31 3:16-19 11:13-16
```

And my webpack config is:
```
const HtmlWebpackPlugin = require("html-webpack-plugin");
const ModuleFederationPlugin = require("webpack").container
  .ModuleFederationPlugin;
const deps = require("./package.json").dependencies;
module.exports = {
  entry: "./src/index",
  cache: false,

  mode: "development",
  devtool: "source-map",

  optimization: {
    minimize: false,
  },

  output: {
    publicPath: "auto",
  },

  resolve: {
    extensions: [".jsx", ".js", ".json",".mjs"],
  },

  module: {
    rules: [
      { test: /\.m?js$/, type: "javascript/auto" },
      {
        test: /\.jsx?$/,
        loader: require.resolve("babel-loader"),
        exclude: /node_modules/,
        options: {
          presets: [require.resolve("@babel/preset-react")],
        },
      },
    ],
  },

  plugins: [
    new ModuleFederationPlugin({
      name: "app_02",
      filename: "remoteEntry.js",
      remotes: {
        app_01: "app_01@http://localhost:3001/remoteEntry.js",
        app_03: "app_03@http://localhost:3003/remoteEntry.js",
      },
      exposes: {
        "./Dialog": "./src/Dialog",
        "./Tabs": "./src/Tabs",
      },
      shared: {
        ...deps,
        "@material-ui/core": {
          singleton: true,
        },
        "react-router-dom": {
          singleton: true,
        },
        "react-dom": {
          singleton: true,
        },
        react: {
          singleton: true,
        },
      },
    }),
    new HtmlWebpackPlugin({
      template: "./public/index.html",
      chunks: ["main"],
    }),
  ],
};
```

### PR DESCRIPTION
**Pull request by non25**
It drops webpack-virtual-modules in favor of much more simple and sane setup described here: https://github.com/webpack/webpack/issues/11074#issuecomment-648658214

The setup works both for webpack 4 and 5. Emitted css reloads just fine on component change on both of them.

This PR changes make use of custom loader string, `virtualModules Map` which contain pairs `*.svelte.{index}.css: content` and `index` to prevent key overlaps when compiling concurrently for SSR and browser.

The loader saves css string to the `virtualModules Map` and appends an import with custom loader string, relaunching svelte-loader with cssPath parameter, which makes it output css basically from the void (without a file) and pass it further to the `.css` loaders chain.

Every time css is read from the `virtualModules Map` it gets deleted to prevent leaks.

I'm currently using this approach in `svelte-loader-hot` on production.

I didn't fix `emitCss: true` test, would be cool if you can help me out.

```
  1) loader
       configuration via query
         emitCss
           should configure emitCss=true:
     AssertionError: expected [Error: TypeError: TypeError: Cannot read property 'replace' of undefined] to not exist
```

Here's what I did
```diff javascript
diff --git a/test/loader.spec.js b/test/loader.spec.js
index e5366a9..e9c40b4 100644
--- a/test/loader.spec.js
+++ b/test/loader.spec.js
@@ -216,7 +216,7 @@ describe('loader', () => {
                                        function(err, code, map) {
                                                expect(err).not.to.exist;
 
-                                               expect(code).to.match(/import '.+\.css';/);
+                                               expect(code).to.match(/!=!svelte-loader\?cssPath=/);
                                        },
                                        { emitCss: true }
                                )
```

Fixes #139, Fixes #131, Fixes #126

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -161,17 +162,11 @@ module.exports = function(source, map) {
 		}
 
 		if (options.emitCss && css.code) {
-			const cssFilepath = compileOptions.filename.replace(
-				/\.[^/.]+$/,
-				`.svelte.css`
-			);
-
+			const resource = posixify(this.resource);
+			const cssPath = `${resource}.${index++}.css`;
```
Comment (Smittyvb) [inline index.js:166]: Perhaps `svelte` should be included in the path to make it clear where the CSS is coming from?

**Comment by Smittyvb** [comment]
Just tested this and it works great, but only when `hotReload` is disabled. When I make a change in watch mode it's reflected as soon as I reload. I do get errors whenever `hotReload` is enabled though.

File: index.js (hunk 1 above)
Comment (Smittyvb) [reply index.js:166]: Oh, I just tried this I realized that `svelte` gets added due to being part of the filename, oops nevermind

**Comment by non25** [comment]
> Just tested this and it works great, but only when `hotReload` is disabled.

`hotReload` doesn't work on master. This PR's goal is not about fixing a `hotReload`.
If we want `hotReload`, we should port `svelte-loader-hot`'s changes here.

**Review by Smittyvb — APPROVED** [review]

File: test/loader.spec.js [hunk 2]
```diff
@@ -208,19 +208,6 @@ describe('loader', () => {
 					{}
 				)
 			);
-
-			it(
-				'should configure emitCss=true',
```
Comment (benmccann) [inline test/loader.spec.js:213]: how come this test was removed?

other than that, this PR looks good to me

Comment (Smittyvb) [reply test/loader.spec.js:213]: the test fails on this branch for some reason with the `expect(err).not.to.exist` line

Comment (benmccann) [reply test/loader.spec.js:213]: It seems like something we should probably fix? I think I'd be happy to merge this after that

Comment (Smittyvb) [reply test/loader.spec.js:213]: Here's the stack trace of the underlying error:
```
CompileError [ValidationError]: A component cannot have a default export
    at error$1 (/home/smit/svelte-loader/node_modules/svelte/compiler.js:15722:20)
    at Component.error (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27639:10)
    at Component.extract_exports (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27671:19)
    at Object.enter (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27931:52)
    at visit (/home/smit/svelte-loader/node_modules/svelte/compiler.js:5256:11)
    at visit (/home/smit/svelte-loader/node_modules/svelte/compiler.js:5288:13)
    at walk (/home/smit/svelte-loader/node_modules/svelte/compiler.js:5207:10)
    at Component.post_template_walk (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27916:10)
    at Component.walk_instance_js_post_template (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27898:15)
    at new Component (/home/smit/svelte-loader/node_modules/svelte/compiler.js:27454:15)
    at compile (/home/smit/svelte-loader/node_modules/svelte/compiler.js:28637:24)
```

Comment (Smittyvb) [reply test/loader.spec.js:213]: Debugging: in `const resource = posixify(this.resource)`, `this.resource` is `undefined`

Comment (Smittyvb) [reply test/loader.spec.js:213]: @non25: I've figured out the issue, https://github.com/non25/svelte-loader/pull/2 re-adds the test in a way that makes it pass.

Comment (non25) [reply test/loader.spec.js:213]: I can't fix it and wrote about it in the post. Would be nice if somebody can help me with that.

File: test/loader.spec.js [hunk 3]
```diff
@@ -208,20 +211,20 @@ describe('loader', () => {
...
+		});
 
-			it(
-				'should configure emitCss=true',
-				testLoader(
-					'test/fixtures/css.html',
-					function(err, code, map) {
-						expect(err).not.to.exist;
+		it(
+			'should configure emitCss=true',
```
Comment (benmccann) [inline test/loader.spec.js:217]: was this moved outside the above function on purpose? does it still work if we put it back where it was originally located?

Comment (non25) [reply test/loader.spec.js]: > was this moved outside the above function on purpose? does it still work if we put it back where it was originally located?

no, fixed it.

**Comment by non25** [comment]
I think I will add one small change and we are good to go.

**Review by benmccann — APPROVED** [review]

**Comment by non25** [comment]
I'll make one slight change.

**Comment by non25** [comment]
I've used compileOptions.filename instead of this.resource, because they are identical. Thus, other edits to tests are not required.

**Comment by newCodeRunner** [comment]
@non25 , can you please share your webpack config regarding this? emitCss: true, always fails for me irrespective of Prod or Dev environment. I followed your multiple PRs but not sure what is going wrong. Thanks in advance,

Here is my webpack config

```
    {
      test: /\.(svelte)$/,
      use: {
        loader: 'svelte-loader',
        options: {
          emitCss: true,
        },
      },
    },
    {
      test: /node_modules\/svelte\/.*\.mjs$/,
      resolve: {
        fullySpecified: false,
      },
    },
    {
      test: /\.css$/,
      use: [
        {
          loader: MiniCssExtractPlugin.loader,
        },
        {
          loader: 'css-loader',
          options: {
            url: false,
          },
        },
     ],
  },
```

**Comment by non25** [comment]
> TypeError: Cannot read property 'data' of undefined
> at  \node_modules\svelte-loader\index.js:181:12

It looks like you are still using published `svelte-loader` in npm.
New changes are not there yet.

Install `svelte-loader` from current master branch:

```bash
# yarn 2
yarn add -D svelte-loader@sveltejs/svelte-loader
# npm
npm install -D sveltejs/svelte-loader
```

Also you can look at https://github.com/non25/webpack-svelte-tests for an inspiration.

**Comment by newCodeRunner** [comment]
@non25 , thank you for your quick response, I will test it now!

**Comment by rwoodnz** [comment]
I used the sveltejs/svelte-loader and the issues I was having went away. However, I am using Github Actions as a CI tool and it gives me "private repositor" build error message. I guess I am waiting for this to publish to NPM then, or is there a way to use this sort of repository from Github Actions? In the meantime I'm using the svelte-loader-hot and that's working ok.

**Comment by kyrylkov** [comment]
@rwoodnz v3 with webpack 5 support is published to npm already https://www.npmjs.com/package/svelte-loader

**Comment by rwoodnz** [comment]
> https://www.npmjs.com/package/svelte-loader

Thanks! I will give it another go

**Comment by rwoodnz** [comment]
That works for me - good work! Should the install instructions save to dev dependencies instead?

**Comment by kyrylkov** [comment]
> Should the install instructions save to dev dependencies instead?

Yes: `npm install -D svelte svelte-loader`

**Pull request merged**